### PR TITLE
AM-57-alternate

### PR DESCRIPTION
**Pull request by farmborough**
## :id: 
Fixes [AM-57](https://graviteecommunity.atlassian.net/browse/AM-57)

## :pencil2: 

Checking passwords against a user's history via the gateway reset form requires communication with the `PasswordHistoryService`. To that end, I've added a new gateway handler and some calling code in `reset_password.html`.

## :computer: 
Password already in history. Can't submit:

Screenshot 2022-10-26 at 10 47 23
<br>
Password not in history, submit ok:
Screenshot 2022-10-26 at 10 47 50

### CODE REVIEW

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/RootProvider.java
Comment (leleueri) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/RootProvider.java:531]: IHMO, we should add the `clientRequestParseHandlerOptional` before calling the PasswordHistoryHandler in order to have access to the client during the password validation. That will allow us to use the right PasswordSettings.

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/handler/user/password/PasswordHistoryHandler.java
Comment (leleueri) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/handler/user/password/PasswordHistoryHandler.java:59]: We should use the PasswordSettings.getSettings method to gety the right password policy rules (app or domain)

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/handler/user/password/PasswordHistoryHandler.java
Comment (leleueri) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/handler/user/password/PasswordHistoryHandler.java:62]: We should use the PasswordSettings.getSettings method to gety the right password policy rules (app or domain)

**Review by leleueri — CHANGES_REQUESTED** [review]

**Review by leleueri — CHANGES_REQUESTED** [review]
The password-validation.js and the reset_password templates should be duplicates in the UI (for the JS file) and the ManagementRestAPI (for the form) in order to have the same version in the preview mode

File: gravitee-am-gateway/gravitee-am-gateway-standalone/gravitee-am-gateway-standalone-distribution/src/main/resources/templates/i18n/messages_fr.properties
Comment (leleueri) [inline gravitee-am-gateway/gravitee-am-gateway-standalone/gravitee-am-gateway-standalone-distribution/src/main/resources/templates/i18n/messages_fr.properties:76]:
```suggestion
password.history.before=Le mot de passe ne doit pas correspondre aux 
```

**Review by leleueri — CHANGES_REQUESTED** [review]

**Review by leleueri — COMMENTED** [review]
Looks good, for the UI part. 
Will approved this PR once the business logic one will be approved.

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/handler/user/password/PasswordHistoryHandler.java
Comment (leleueri) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/handler/user/password/PasswordHistoryHandler.java:57]: For safety, I think we should manage the error case into this handler. Otherwise, if the `passwordAlreadyUsed` throw an exception or return a Single.error then the GW will never respond

**Review by leleueri — CHANGES_REQUESTED** [review]

**Comment by farmborough** [comment]
> SonarCloud Quality Gate failed.    [Quality Gate failed](https://sonarcloud.io/dashboard?id=gravitee-io_graviteeio-access-management&pullRequest=2202)
> 
> [Bug](https://sonarcloud.io/project/issues?id=gravitee-io_graviteeio-access-management&pullRequest=2202&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=gravitee-io_graviteeio-access-management&pullRequest=2202&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=gravitee-io_graviteeio-access-management&pullRequest=2202&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=gravitee-io_graviteeio-access-management&pullRequest=2202&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=gravitee-io_graviteeio-access-management&pullRequest=2202&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=gravitee-io_graviteeio-access-management&pullRequest=2202&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=gravitee-io_graviteeio-access-management&pullRequest=2202&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=gravitee-io_graviteeio-access-management&pullRequest=2202&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=gravitee-io_graviteeio-access-management&pullRequest=2202&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=gravitee-io_graviteeio-access-management&pullRequest=2202&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=gravitee-io_graviteeio-access-management&pullRequest=2202&resolved=false&types=CODE_SMELL) [0 Code Smells](https://sonarcloud.io/project/issues?id=gravitee-io_graviteeio-access-management&pullRequest=2202&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=gravitee-io_graviteeio-access-management&pullRequest=2202&metric=new_coverage&view=list) [0.0% Coverage](https://sonarcloud.io/component_measures?id=gravitee-io_graviteeio-access-management&pullRequest=2202&metric=new_coverage&view=list) [](https://sonarcloud.io/component_measures?id=gravitee-io_graviteeio-access-management&pullRequest=2202&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=gravitee-io_graviteeio-access-management&pullRequest=2202&metric=new_duplicated_lines_density&view=list)

Sonar still not recognising Junit 5 tests.  🤬

**Review by leleueri — APPROVED** [review]

**Pull request merged**
